core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1048,7 +1065,7 @@ private[spark] class BlockManager(
           try {
             replicate(blockId, bytesToReplicate, level, remoteClassTag)
           } finally {
-            bytesToReplicate.dispose()
+            bytesToReplicate.unmap()
```

Comment:
I would to explore this further to ensure this would really work well, but I like your idea with one caveat. I think we should avoid using the `disposed` var for this purpose. Using it this way would introduce ambiguity in its meaning when `disposed` is true. In addition to its current meaning, it could also mean "this buffer is not disposed but we don't want to dispose it".

Instead, I suggest keeping the usage of `disposed` as-is and adding an additional var, e.g. `indisposable`, which makes the `dispose` method itself a no-op.